### ISSUE TITLE
Object name: 'System.Net.Sockets.Socket'.) Cannot access a disposed object.

### ISSUE THREAD

**Issue by jinek** [dotnet/SqlClient#814]
### Describe the bug

Exception arrives to TaskScheduler.UnobservedTaskException.

```
_Exception message:_ 
A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.) Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.

_Stack trace:_
System.AggregateException:
System.ObjectDisposedException:
   at System.Net.Sockets.Socket.EndReceiveFrom (System.Net.Sockets, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.Net.Sockets.UdpClient.EndReceive (System.Net.Sockets, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.Net.Sockets.UdpClient+<>c.<ReceiveAsync>b__57_1 (System.Net.Sockets, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
```

### Expected behavior
No exceptions arrive to TaskScheduler.UnobservedTaskException.

### Further technical details
Host (useful for support):
Version: 3.0.3
Commit: c03f2fe626

.NET Core SDKs installed:
No SDKs were found.

.NET Core runtimes installed:
Microsoft.AspNetCore.App 3.0.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.NETCore.App 3.0.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

**Repro**
```
            int unobservedCount=0;
            // possible to see the issue with debugger attached or without
            TaskScheduler.UnobservedTaskException += (sender, eventArgs) =>
            {
                eventArgs!.Exception!.Handle(exception =>
                {
                    Debug.WriteLine($"Unobserved exception {eventArgs.Exception}");
                    Interlocked.Increment(ref unobservedCount);
                    return true; // for debugging purposes ( counting unobserved exceptions )
                });
            };
            
            // pretending to be SqlBrowser 
            UdpClient udpClient = new UdpClient();
            udpClient.Client.Bind(new IPEndPoint(IPAddress.Any, 1434));

            using SqlConnection connection = new SqlConnection(@"Data Source=127.0.0.1;User ID=sa;Password=sa;Initial Catalog=sa;");

            try
            {
                connection.Open(); // internally ObjectDisposedException happens each second
            }
            catch (SqlException sqlException)
            {
                GC.Collect(0);
                GC.WaitForPendingFinalizers();
                
                // With Connection timeout 15, should be about 15 unobserved exceptions
                
                // Also sqlException contains code 35 (provider: TCP Provider, error: 35 - An internal exception was caught)
                if(unobservedCount>0)
                    throw new ApplicationException($"Number of unobserved exceptions {unobservedCount}", sqlException);
            }
```

**Additional context**
@cheenamalhotra Hello. Seems this issue is very similar to https://github.com/dotnet/SqlClient/issues/799 and is caused by leaving tasks in unobserved states.
It seems this issue is reproducable in all versions where SSRP.cs is used:

I believe this line is the issue:
https://github.com/dotnet/SqlClient/blob/ac8bb3f9132e6c104dc3e307fe2d569daed0776f/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SSRP.cs#L168
In case of timeout happened it leaves one of the task in unobserved state, same time implicitely calling the **UdpClient.Dispose()**, which later raises **ObjectDisposedException** from **UdpClient** in that unobserved task.

Still, I think it makes sense to wait for the end of discussion here: https://github.com/dotnet/runtime/issues/44770#issuecomment-731564910 I have assumed, that may be OperationCancelledException should be raised instead of ObjectDisposedException in this situation. If that is the case, then this issue would be fixed.

Additionally, may I ask small help for understand those timeouts. They are fixed on one second for UdpClient.SendAsync and one second for UdpClient.ReceiveAsync. For me it seems strange that those timeouts are not anyhow dependent on SqlConnection timeouts. Lets say, with this conditions a user can "randomly" receive connection error or probably protocol/ip-switching with no visible reason.

**Comment by jinek** [comment]
@cheenamalhotra  So the general idea is about how fast is reponse from SqlBrowser. If it's less then 1 second, it comes to confusing message:  
```
" ---> System.Data.SqlClient.SqlException: A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: TCP Provider, error: 35 - An internal exception was caught)
 ---> System.AggregateException: One or more errors occurred. (Connection refused 127.0.0.1:1433)"
```
In fact, it also failed to receive response from 1434 in time.
Those exceptions are ignored by defaults of the runtime, I think this is another issue, still may be related to issues similar to this: https://github.com/dotnet/SqlClient/issues/75

As I've mentioned, for me it also seems strange, that timeouts are set to 1 second constantly. Because if we have 15 seconds overall timeout, there is better chance to receive SqlBrowser answer within up to 15 seconds timeout.

**Comment by cheenamalhotra** [comment]
Hi @jinek 

Could you provide link/upload your repro here?
I'll have to dig in the design as it's an old implementation, before my time :)

**Comment by jinek** [comment]
> Could you provide link/upload your repro here?
> I'll have to dig in the design as it's an old implementation, before my time :)

I mean it's at the top of this page)

Or if you can not reach the top of this page, here it is)))

```
  int unobservedCount=0;
            // possible to see the issue with debugger attached or without
            TaskScheduler.UnobservedTaskException += (sender, eventArgs) =>
            {
                eventArgs!.Exception!.Handle(exception =>
                {
                    Debug.WriteLine($"Unobserved exception {eventArgs.Exception}");
                    Interlocked.Increment(ref unobservedCount);
                    return true; // for debugging purposes ( counting unobserved exceptions )
                });
            };
            
            // pretending to be SqlBrowser 
            UdpClient udpClient = new UdpClient();
            udpClient.Client.Bind(new IPEndPoint(IPAddress.Any, 1434));

            using SqlConnection connection = new SqlConnection(@"Data Source=127.0.0.1;User ID=sa;Password=sa;Initial Catalog=sa;");

            try
            {
                connection.Open(); // internally ObjectDisposedException happens each second
            }
            catch (SqlException sqlException)
            {
                GC.Collect(0);
                GC.WaitForPendingFinalizers();
                
                // With Connection timeout 15, should be about 15 unobserved exceptions
                
                // Also sqlException contains code 35 (provider: TCP Provider, error: 35 - An internal exception was caught)
                if(unobservedCount>0)
                    throw new ApplicationException($"Number of unobserved exceptions {unobservedCount}", sqlException);
            }
```

**Comment by cheenamalhotra** [comment]
Duh!
My bad, I scrolled too fast 🙈 
Thx :)

**Comment by cheenamalhotra** [comment]
I can't manage to run this:
```
udpClient.Client.Bind(new IPEndPoint(IPAddress.Any, 1434));
```
As I get error:

`System.Net.Sockets.SocketException: 'Only one usage of each socket address (protocol/network address/port) is normally permitted.'`

How do you get past this?

**Comment by jinek** [comment]
> `System.Net.Sockets.SocketException: 'Only one usage of each socket address (protocol/network address/port) is normally permitted.'`
> 
> How do you get past this?

Seems you have Sql installed on same machine. Thus, SqlBrowser keeps this port open: https://docs.microsoft.com/en-us/sql/tools/configuration-manager/sql-server-browser-sql-server-configuration-manager?view=sql-server-ver15

But this issue is claimed to be reproduced only on linux machine.

**Comment by udlose** [comment]
Is there any update on this issue?

**Comment by nikhilgoud** [comment]
Any update on this ? 
I've got this error in AWS Lambda 

`
ErrorMessage : Cannot access a disposed object. Object name: 'System.Net.Sockets.Socket'. . Details: . at System.Net.Sockets.Socket.SetSocketOption(SocketOptionLevel optionLevel, SocketOptionName optionName, Int32 optionValue) at System.Data.SqlClient.SNI.SNITCPHandle.Receive(SNIPacket& packet, Int32 timeoutInMilliseconds) at System.Data.SqlClient.SNI.SNIProxy.ReadSyncOverAsync(SNIHandle handle, SNIPacket& packet, Int32 timeout) at System.Data.SqlClient.SNI.TdsParserStateObjectManaged.ReadSyncOverAsync(Int32 timeoutRemaining, UInt32& error) at System.Data.SqlClient.TdsParserStateObject.ReadSniSyncOverAsync() at System.Data.SqlClient.TdsParserStateObject.TryReadNetworkPacket() at System.Data.SqlClient.TdsParser.ConsumePreLoginHandshake(Boolean encrypt, Boolean trustServerCert, Boolean integratedSecurity, Boolean& marsCapable, Boolean& fedAuthRequired) at System.Data.SqlClient.TdsParser.Connect(ServerInfo serverInfo, SqlInternalConnectionTds connHandler, Boolean ignoreSniOpenTimeout, Int64 timerExpire, Boolean encrypt, Boolean trustServerCert, Boolean integratedSecurity, Boolean withFailover) at System.Data.SqlClient.SqlInternalConnectionTds.AttemptOneLogin(ServerInfo serverInfo, String newPassword, SecureString newSecurePassword, Boolean ignoreSniOpenTimeout, TimeoutTimer timeout, Boolean withFailover) at System.Data.SqlClient.SqlInternalConnectionTds.LoginNoFailover(ServerInfo serverInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString connectionOptions, SqlCredential credential, TimeoutTimer timeout) at System.Data.SqlClient.SqlInternalConnectionTds.OpenLoginEnlist(TimeoutTimer timeout, SqlConnectionString connectionOptions, SqlCredential credential, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance) at System.Data.SqlClient.SqlInternalConnectionTds..ctor(DbConnectionPoolIdentity identity, SqlConnectionString connectionOptions, SqlCredential credential, Object providerInfo, String newPassword, SecureString newSecurePassword, Boolean redirectedUserInstance, SqlConnectionString userConnectionOptions, SessionData reconnectSessionData, Boolean applyTransientFaultHandling, String accessToken) at System.Data.SqlClient.SqlConnectionFactory.CreateConnection(DbConnectionOptions options, DbConnectionPoolKey poolKey, Object poolGroupProviderInfo, DbConnectionPool pool, DbConnection owningConnection, DbConnectionOptions userOptions) at System.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(DbConnectionPool pool, DbConnection owningObject, DbConnectionOptions options, DbConnectionPoolKey poolKey, DbConnectionOptions userOptions) at System.Data.ProviderBase.DbConnectionPool.CreateObject(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection) at System.Data.ProviderBase.DbConnectionPool.UserCreateRequest(DbConnection owningObject, DbConnectionOptions userOptions, DbConnectionInternal oldConnection) at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection) at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection) at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection) at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions) at System.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions) at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry) at System.Data.SqlClient.SqlConnection.Open() at Dapper.SqlMapper.QueryImpl[T](IDbConnection cnn, CommandDefinition command, Type effectiveType)+MoveNext() in /_/Dapper/SqlMapper.cs:line 1083 at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection) at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source) at Dapper.SqlMapper.Query[T](IDbConnection cnn, String sql, Object param, IDbTransaction transaction, Boolean buffered, Nullable`1 commandTimeout, Nullable`1 commandType) in /_/Dapper/SqlMapper.cs:line 725 .....
....`

**Comment by JRahnama** [comment]
@nikhilgoud based on the trace you are using System.Data.SqlClient and that library is in servicing mode and is not taking any fix. Can you try with Microsoft.Data.SqlClient please?

**Comment by Mek7** [comment]
I am currently migrating a large application from .net framework to .net 8. So far, the application still uses the old EF 6, with EF Core planned when everything is migrated to .net 8.
As part of this process, some https endpoints were also containerized for linux and are hosted in Azure Container Apps.
After a few days of higher load, our entire application starts to fail with the exception mentioned above. We have a retry policy that handles this exception, but retry never succeeds because the whole connection pool seems to be corrupted.
Only restart of whole container app helps.
So far I implemented an IDbCommandInterceptor for EF datacontext that clears the pool when this exception occurs.

System.Data.SqlClient is used because it is native to old EF. Migrating to EF Core to use the Microsoft.Data.SqlClient is a looooong way.
Can't this System.Data.SqlClient be fixed once and for all? It's a serious problem that more and more people will stumble upon so I think it would be good to have a fix even in the "servicing mode".

**Comment by ErikEJ** [comment]
@Mek7 why not use Microsoft.Entityframework.SqlServer ?

**Comment by Mek7** [comment]
I didn't know about it... but I tried it and it worked and solved this error.
For anyone else - there is a nice guide in MS docs what you have to do to use Microsoft.Entityframework.SqlServer
Thank you for your suggestion.

**Comment by ErikEJ** [comment]
@Mek7 Thanks, I wrote that guide!

**Comment by jinek** [comment]
@cheenamalhotra this one was fixed
